### PR TITLE
fix(melange): add js rules for vlib impl modules

### PR DESCRIPTION
**Pull request by jchavarri**
Js rules for impl modules in vlibs were not being created.

This path updates the tests to cover that case, plus adapts `Melange_rules` so that when adding rules for a library, the `implements` field is checked in order to add rules for the vlib as well.

The decision is to leave vlib js artifacts inside same vlib folder, so `dst_dir` is modified for vlibs to point to the right location.

### CODE REVIEW

File: src/dune_rules/melange_rules.ml [hunk 1]
```diff
@@ -9,6 +9,33 @@ let make_js_name ~js_ext ~dst_dir m =
   let name = Melange.js_basename m ^ js_ext in
   Path.Build.relative dst_dir name
 
+let virtual_dst_dir_and_modules sctx ~target_dir ~implements ~scope =
```
Comment (jchavarri) [inline src/dune_rules/melange_rules.ml:12]: I based this function on `Virtual_rules.impl`, but it doesn't compute anything related to `foreign_objects`, and it retrieves the `dst_dir` information.

Not sure if there are opportunities to consolidate, it seemed cleaner to have a separate implementation, considering how specific Melange is.

File: src/dune_rules/melange_rules.ml [hunk 2]
```diff
@@ -29,21 +56,38 @@ let js_includes ~loc ~sctx ~target_dir ~requires_link ~scope ~js_ext =
...
-          let* source_modules = modules_group >>| Modules.impl_only in
-          let of_module m =
-            let output = make_js_name ~js_ext ~dst_dir m in
-            Dep.file (Path.build output)
-          in
-          Resolve.Memo.return
-            (List.map source_modules ~f:of_module |> Dep.Set.of_list)
+          match Lib_info.virtual_ info with
+          | Some _ ->
+            (* Js rules for virtual libs are added by their implementations *)
```
Comment (jchavarri) [inline src/dune_rules/melange_rules.ml:61]: This was surprising, but it turns out that when calculating `Lib.Compile.requires_link` for an implementation library in `add_rules_for_libraries`

https://github.com/ocaml/dune/blob/ef472e94e3ab76a4f2e63d1d7043f77cec5319ab/src/dune_rules/melange_rules.ml#L194-L195

the virtual library is returned as part of the list in the resulting value 🤔

File: test/blackbox-tests/test-cases/melange/virtual_lib.t/run.t [hunk 3]
```diff
@@ -1,5 +1,12 @@
...
+Virtual library with a single module are supported by melange libs
 
   $ dune build output/c.js
   $ node _build/default/output/c.js
   hello from melange
+
+Implementation modules in the virtual library are supported
+
+  $ touch vlib/shared.ml
+  $ dune build output/c.js
```
Comment (anmonteiro) [inline test/blackbox-tests/test-cases/melange/virtual_lib.t/run.t:10]: was this returning an error before? can you `ls _build/default/output` to check that `shared.js` ends up there?

Comment (jchavarri) [reply test/blackbox-tests/test-cases/melange/virtual_lib.t/run.t:10]: Yes, see https://github.com/ocaml/dune/pull/6495/commits/72769c8471685b972c7bb0128611f7ddf0b30831.

> can you ls _build/default/output to check that shared.js ends up there?

Even more, I am thinking about updating the test to make it something realistic: have exe / melange implementations, intra-deps in the vlib, in the impl lib, etc. I think some cases are still broken, due to `obj_dir` of vlib not being taken into account.

Comment (jchavarri) [reply test/blackbox-tests/test-cases/melange/virtual_lib.t/run.t]: added more complete test in https://github.com/ocaml/dune/pull/6495/commits/f808e3bc50a854c22f057c007e56bc2561bf797f.

File: src/dune_rules/melange_rules.ml (hunk 2 above)
Comment (rgrinberg) [reply src/dune_rules/melange_rules.ml:61]: That sounds wrong. Virtual libraries are supposed to be absent from the linking closure.

File: src/dune_rules/melange_rules.ml [hunk 4]
```diff
@@ -9,6 +9,33 @@ let make_js_name ~js_ext ~dst_dir m =
...
   Path.Build.relative dst_dir name
 
+let virtual_dst_dir_and_modules sctx ~target_dir ~implements ~scope =
+  let open Memo.O in
+  match implements with
+  | None -> Memo.return None
+  | Some (loc, implements) -> (
+    Lib.DB.find (Scope.libs scope) implements >>= function
+    | None ->
+      User_error.raise ~loc
```
Comment (rgrinberg) [inline src/dune_rules/melange_rules.ml:19]: This also breaks loading the rules, but it's fine for now

File: src/dune_rules/melange_rules.ml [hunk 5]
```diff
@@ -181,6 +225,7 @@ let add_rules_for_libraries ~dir ~scope ~target_dir ~sctx ~requires_link
       in
       let lib = local_of_lib ~loc:mel.loc lib in
       let info = Lib.Local.info lib in
+      let implements = Lib_info.implements info in
```
Comment (rgrinberg) [inline src/dune_rules/melange_rules.ml:228]: There's Lib.implements that does resolution for you

File: src/dune_rules/melange_rules.ml (hunk 2 above)
Comment (jchavarri) [reply src/dune_rules/melange_rules.ml:61]: I took a look, but I am not sure how to create a test that proofs this. Do you have any ideas, or existing tests that I could use as reference?

File: src/dune_rules/melange_rules.ml (hunk 4 above)
Comment (jchavarri) [reply src/dune_rules/melange_rules.ml:19]: Is there a rule of thumb to know when it's fine to raise user errors and break the rules loading? I have the impression that my criteria is generally more lenient (e.g. #6412 was fine to fail early), but I don't get the big picture yet.

File: src/dune_rules/melange_rules.ml (hunk 5 above)
Comment (jchavarri) [reply src/dune_rules/melange_rules.ml]: Using it in https://github.com/ocaml/dune/pull/6495/commits/fb4874199db5ec5d0f3df77485d8ab60c2132874.

File: src/dune_rules/melange_rules.ml [hunk 6]
```diff
@@ -9,6 +9,28 @@ let make_js_name ~js_ext ~dst_dir m =
...
+      let dir = Lib_info.src_dir info in
+      Dir_contents.get sctx ~dir
+    in
+    let* modules =
+      Dir_contents.ocaml dir_contents
+      >>| Ml_sources.modules ~for_:(Library name)
+    in
+    let lib_dir = Lib_info.src_dir info in
+    let dst_dir = lib_output_dir ~target_dir ~lib_dir in
+    Memo.return (Some (dst_dir, modules))
```
Comment (rgrinberg) [inline src/dune_rules/melange_rules.ml:32]: There's Dir_contents.modules_of_lib to simplify this.

File: src/dune_rules/melange_rules.ml (hunk 2 above)
Comment (rgrinberg) [reply src/dune_rules/melange_rules.ml:58]: No way to write a test but you can add an assert inside the function that calculates the closure.

File: src/dune_rules/melange_rules.ml [hunk 7]
```diff
@@ -27,23 +49,43 @@ let js_includes ~loc ~sctx ~target_dir ~requires_link ~scope ~js_ext =
...
+          | None ->
+            let lib_dir = Lib_info.src_dir info in
+            let dst_dir = lib_output_dir ~target_dir ~lib_dir in
+            let open Memo.O in
+            let modules_group =
+              Dir_contents.get sctx ~dir:lib_dir
+              >>= Dir_contents.ocaml
+              >>| Ml_sources.modules ~for_:(Library lib_name)
+            in
+            let* source_modules = modules_group >>| Modules.impl_only in
```
Comment (rgrinberg) [inline src/dune_rules/melange_rules.ml:69]: This isn't quite right I believe. `impl_only` returns all the modules with implementations which will include modules from  the vlib and impl. This means you'll set up the same rules twice. This is how it should go:

1. Get the modules for a library using `Dir_contents.modules_of_lib`.
2. Use `Modules.split_by_lib` to get the modules for the virtual library and implementation separately
3. Setup rules for every individual module if `Module.has ~ml_kind:Impl` returns true.

**Comment by rgrinberg** [comment]
@jchavarri do you mind if I finish this PR for you?

**Comment by jchavarri** [comment]
Sure, no problem. The PR was de-prioritized after making the decision to
use copy_files for now.

File: src/dune_rules/melange_rules.ml (hunk 6 above)
Comment (jchavarri) [reply src/dune_rules/melange_rules.ml]: https://github.com/ocaml/dune/pull/6495/commits/f9b9ab21f2c82ddfd533fd0ef0cbdc1514209884

**Comment by jchavarri** [comment]
Superseded by https://github.com/ocaml/dune/pull/6693.

**Pull request closed without merging**